Patch by MSP-Greg
Renames duplicate class names that cause intermittent loss of tests when
test-all is run with a -j parameter
diff --git a/test/-ext-/array/test_resize.rb b/test/-ext-/array/test_resize.rb
index f6a368cb75..c7a57f3cf1 100644
--- a/test/-ext-/array/test_resize.rb
+++ b/test/-ext-/array/test_resize.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require '-test-/array/resize'
 
-class TestArray < Test::Unit::TestCase
+class Test_Array < Test::Unit::TestCase
   class TestResize < Test::Unit::TestCase
     def test_expand
       feature = '[ruby-dev:42912]'
diff --git a/test/-ext-/bignum/test_big2str.rb b/test/-ext-/bignum/test_big2str.rb
index 296a5a9c49..cc28d97ce5 100644
--- a/test/-ext-/bignum/test_big2str.rb
+++ b/test/-ext-/bignum/test_big2str.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require "-test-/bignum"
 
-class TestBignum < Test::Unit::TestCase
+class Test_Bignum < Test::Unit::TestCase
   class TestBig2str < Test::Unit::TestCase
 
     SIZEOF_BDIGIT = Integer::SIZEOF_BDIGIT
diff --git a/test/-ext-/bignum/test_bigzero.rb b/test/-ext-/bignum/test_bigzero.rb
index 43a99b01a3..cf34964acd 100644
--- a/test/-ext-/bignum/test_bigzero.rb
+++ b/test/-ext-/bignum/test_bigzero.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require "-test-/bignum"
 
-class TestBignum < Test::Unit::TestCase
+class Test_Bignum < Test::Unit::TestCase
   class TestBigZero < Test::Unit::TestCase
     def test_equal_0
       bug8204 = '[ruby-core:53893] [Bug #8204]'
diff --git a/test/-ext-/bignum/test_div.rb b/test/-ext-/bignum/test_div.rb
index c53332630a..4e4bd3fa3e 100644
--- a/test/-ext-/bignum/test_div.rb
+++ b/test/-ext-/bignum/test_div.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require "-test-/bignum"
 
-class TestBignum < Test::Unit::TestCase
+class Test_Bignum < Test::Unit::TestCase
   class TestDiv < Test::Unit::TestCase
 
     SIZEOF_BDIGIT = Integer::SIZEOF_BDIGIT
diff --git a/test/-ext-/bignum/test_mul.rb b/test/-ext-/bignum/test_mul.rb
index fa974ab55d..5de046a804 100644
--- a/test/-ext-/bignum/test_mul.rb
+++ b/test/-ext-/bignum/test_mul.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require "-test-/bignum"
 
-class TestBignum < Test::Unit::TestCase
+class Test_Bignum < Test::Unit::TestCase
   class TestMul < Test::Unit::TestCase
 
     SIZEOF_BDIGIT = Integer::SIZEOF_BDIGIT
diff --git a/test/-ext-/bignum/test_pack.rb b/test/-ext-/bignum/test_pack.rb
index 04bf3e02de..5d553435fe 100644
--- a/test/-ext-/bignum/test_pack.rb
+++ b/test/-ext-/bignum/test_pack.rb
@@ -4,7 +4,7 @@
 require 'test/unit'
 require "-test-/bignum"
 
-class TestBignum < Test::Unit::TestCase
+class Test_Bignum < Test::Unit::TestCase
   class TestPack < Test::Unit::TestCase
 
     MSWORD_FIRST = Integer::INTEGER_PACK_MSWORD_FIRST
diff --git a/test/-ext-/bignum/test_str2big.rb b/test/-ext-/bignum/test_str2big.rb
index c14811d9c4..6886e4a968 100644
--- a/test/-ext-/bignum/test_str2big.rb
+++ b/test/-ext-/bignum/test_str2big.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require "-test-/bignum"
 
-class TestBignum < Test::Unit::TestCase
+class Test_Bignum < Test::Unit::TestCase
   class TestStr2big < Test::Unit::TestCase
 
     SIZEOF_BDIGIT = Integer::SIZEOF_BDIGIT
diff --git a/test/-ext-/exception/test_enc_raise.rb b/test/-ext-/exception/test_enc_raise.rb
index 2bc7f02413..1faa500b25 100644
--- a/test/-ext-/exception/test_enc_raise.rb
+++ b/test/-ext-/exception/test_enc_raise.rb
@@ -3,7 +3,7 @@
 require '-test-/exception'
 
 module Bug
-  class TestException < Test::Unit::TestCase
+  class Test_Exception < Test::Unit::TestCase
     def test_enc_raise
       feature5650 = '[ruby-core:41160]'
       Encoding.list.each do |enc|
diff --git a/test/-ext-/hash/test_delete.rb b/test/-ext-/hash/test_delete.rb
index e2ad3cbdbc..91e6ff67cb 100644
--- a/test/-ext-/hash/test_delete.rb
+++ b/test/-ext-/hash/test_delete.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require '-test-/hash'
 
-class TestHash < Test::Unit::TestCase
+class Test_Hash < Test::Unit::TestCase
   class TestDelete < Test::Unit::TestCase
     def test_delete
       hash = Bug::Hash.new
diff --git a/test/-ext-/integer/test_integer.rb b/test/-ext-/integer/test_integer.rb
index 320237791a..cdabb5403b 100644
--- a/test/-ext-/integer/test_integer.rb
+++ b/test/-ext-/integer/test_integer.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require '-test-/integer'
 
-class TestInteger < Test::Unit::TestCase
+class Test_Integer < Test::Unit::TestCase
   FIXNUM_MIN = RbConfig::LIMITS['FIXNUM_MIN']
   FIXNUM_MAX = RbConfig::LIMITS['FIXNUM_MAX']
 
diff --git a/test/-ext-/integer/test_my_integer.rb b/test/-ext-/integer/test_my_integer.rb
index 260986d203..1b6f8489f8 100644
--- a/test/-ext-/integer/test_my_integer.rb
+++ b/test/-ext-/integer/test_my_integer.rb
@@ -2,7 +2,7 @@
 require 'test/unit'
 require "-test-/integer"
 
-class TestIntegerExt < Test::Unit::TestCase
+class Test_MyInteger < Test::Unit::TestCase
   def test_my_integer_to_f
     assert_raise(NotImplementedError) do
       Bug::Integer::MyInteger.new.to_f
diff --git a/test/-ext-/method/test_arity.rb b/test/-ext-/method/test_arity.rb
index 2ba73c8d4c..cb8549e7c9 100644
--- a/test/-ext-/method/test_arity.rb
+++ b/test/-ext-/method/test_arity.rb
@@ -2,7 +2,7 @@
 require '-test-/method'
 require 'test/unit'
 
-class TestMethod < Test::Unit::TestCase
+class Test_Method < Test::Unit::TestCase
   class TestArity < Test::Unit::TestCase
     class A
       def foo0()
diff --git a/test/-ext-/proc/test_bmethod.rb b/test/-ext-/proc/test_bmethod.rb
index 344f975755..403d99bfd1 100644
--- a/test/-ext-/proc/test_bmethod.rb
+++ b/test/-ext-/proc/test_bmethod.rb
@@ -2,12 +2,12 @@
 require 'test/unit'
 require '-test-/proc'
 
-class TestProc < Test::Unit::TestCase
+class Test_Proc < Test::Unit::TestCase
   class TestBMethod < Test::Unit::TestCase
   end
 end
 
-class TestProc::TestBMethod
+class Test_Proc::TestBMethod
   class Base
     def foo(*a)
       a
diff --git a/test/-ext-/test_notimplement.rb b/test/-ext-/test_notimplement.rb
index 0eba7bdaf8..80757313ca 100644
--- a/test/-ext-/test_notimplement.rb
+++ b/test/-ext-/test_notimplement.rb
@@ -1,7 +1,7 @@
 # frozen_string_literal: false
 require '-test-/notimplement'
 
-class TestNotImplement < Test::Unit::TestCase
+class Test_NotImplement < Test::Unit::TestCase
   def test_funcall_notimplement
     bug3662 = '[ruby-dev:41953]'
     assert_raise(NotImplementedError, bug3662) {
diff --git a/test/erb/test_erb.rb b/test/erb/test_erb.rb
index 6a6ce3d129..665d1e2933 100644
--- a/test/erb/test_erb.rb
+++ b/test/erb/test_erb.rb
@@ -2,6 +2,7 @@
 # frozen_string_literal: false
 require 'test/unit'
 require 'erb'
+require 'stringio'
 
 class TestERB < Test::Unit::TestCase
   class MyError < RuntimeError ; end
@@ -235,6 +236,7 @@ def test_run
     $stdout = orig
     out.rewind
     assert_equal('9', out.read)
+    return unless num               # to remove warning
   end
 
   class Foo; end
diff --git a/test/erb/test_erb_m17n.rb b/test/erb/test_erb_m17n.rb
index a7840c9605..2fd9e700f3 100644
--- a/test/erb/test_erb_m17n.rb
+++ b/test/erb/test_erb_m17n.rb
@@ -3,7 +3,7 @@
 require 'test/unit'
 require 'erb'
 
-class TestERB < Test::Unit::TestCase
+class TestERBEncoding < Test::Unit::TestCase
   def test_result_encoding
     erb = ERB.new("hello")
     assert_equal __ENCODING__, erb.result.encoding
diff --git a/test/fileutils/test_dryrun.rb b/test/fileutils/test_dryrun.rb
index 97d72fcdf9..fd8a7805ec 100644
--- a/test/fileutils/test_dryrun.rb
+++ b/test/fileutils/test_dryrun.rb
@@ -8,7 +8,7 @@
 class TestFileUtilsDryRun < Test::Unit::TestCase
 
   include FileUtils::DryRun
-  include TestFileUtils::Visibility
+  include TestFileUtilsInc::Visibility
 
   def setup
     super
diff --git a/test/fileutils/test_nowrite.rb b/test/fileutils/test_nowrite.rb
index b804498525..543fa39f5a 100644
--- a/test/fileutils/test_nowrite.rb
+++ b/test/fileutils/test_nowrite.rb
@@ -8,7 +8,7 @@
 class TestFileUtilsNoWrite < Test::Unit::TestCase
 
   include FileUtils::NoWrite
-  include TestFileUtils::Visibility
+  include TestFileUtilsInc::Visibility
 
   def setup
     super
diff --git a/test/fileutils/test_verbose.rb b/test/fileutils/test_verbose.rb
index 0019946393..cf65be8e03 100644
--- a/test/fileutils/test_verbose.rb
+++ b/test/fileutils/test_verbose.rb
@@ -8,7 +8,7 @@
 class TestFileUtilsVerbose < Test::Unit::TestCase
 
   include FileUtils::Verbose
-  include TestFileUtils::Visibility
+  include TestFileUtilsInc::Visibility
 
   def setup
     super
diff --git a/test/fileutils/visibility_tests.rb b/test/fileutils/visibility_tests.rb
index 0ce00d5f2c..4c02c9d207 100644
--- a/test/fileutils/visibility_tests.rb
+++ b/test/fileutils/visibility_tests.rb
@@ -2,14 +2,14 @@
 require 'test/unit'
 require 'fileutils'
 
-class TestFileUtils < Test::Unit::TestCase
+class TestFileUtilsInc < Test::Unit::TestCase
 end
 
 ##
 # These tests are reused in the FileUtils::Verbose, FileUtils::NoWrite and
 # FileUtils::DryRun tests
 
-module TestFileUtils::Visibility
+module TestFileUtilsInc::Visibility
 
   FileUtils::METHODS.each do |m|
     define_method "test_singleton_visibility_#{m}" do
diff --git a/test/optparse/test_zsh_completion.rb b/test/optparse/test_zsh_completion.rb
index 1bc6c7f0bb..8411bc41d4 100644
--- a/test/optparse/test_zsh_completion.rb
+++ b/test/optparse/test_zsh_completion.rb
@@ -4,7 +4,7 @@
 
 class TestOptionParser < Test::Unit::TestCase
 end
-class TestOptionParser::BashCompletion < Test::Unit::TestCase
+class TestOptionParser::BashCompletionZSH < Test::Unit::TestCase
   def setup
     @opt = OptionParser.new
     @opt.define("-z", "zzz") {}
diff --git a/test/rexml/test_martin_fowler.rb b/test/rexml/test_martin_fowler.rb
index da685a80ec..add3c82723 100644
--- a/test/rexml/test_martin_fowler.rb
+++ b/test/rexml/test_martin_fowler.rb
@@ -3,7 +3,7 @@
 require 'rexml/document'
 
 module REXMLTests
-  class OrderTester < Test::Unit::TestCase
+  class OrderTesterMF < Test::Unit::TestCase
     DOC = <<END
 <paper>
 <title>Remove this element and figs order differently</title>
diff --git a/test/ruby/test_ifunless.rb b/test/ruby/test_ifunless.rb
index d533e155bc..f68e5154a2 100644
--- a/test/ruby/test_ifunless.rb
+++ b/test/ruby/test_ifunless.rb
@@ -1,7 +1,7 @@
 # frozen_string_literal: false
 require 'test/unit'
 
-class TestIfunless < Test::Unit::TestCase
+class TestIfUnless < Test::Unit::TestCase
   def test_if_unless
     x = 'test';
     assert(if x == x then true else false end)
diff --git a/test/ruby/test_not.rb b/test/ruby/test_not.rb
index 721f868a5a..12e4c4b696 100644
--- a/test/ruby/test_not.rb
+++ b/test/ruby/test_not.rb
@@ -1,7 +1,7 @@
 # frozen_string_literal: false
 require 'test/unit'
 
-class TestIfunless < Test::Unit::TestCase
+class TestNot < Test::Unit::TestCase
   def test_not_with_grouped_expression
     assert_equal(false, (not (true)))
     assert_equal(true, (not (false)))
diff --git a/test/ruby/test_weakmap.rb b/test/ruby/test_weakmap.rb
index cde186c5f3..35025c8201 100644
--- a/test/ruby/test_weakmap.rb
+++ b/test/ruby/test_weakmap.rb
@@ -39,7 +39,7 @@ def test_include?
       x = nil
     end
     GC.start
-    skip # TODO: failure introduced from r60440
+    # skip('TODO: failure introduced from r60440')
     assert_not_send([@wm, m, k])
   end
   alias test_member? test_include?
diff --git a/test/win32ole/test_thread.rb b/test/win32ole/test_thread.rb
index f1cdfe32b0..cb34693064 100644
--- a/test/win32ole/test_thread.rb
+++ b/test/win32ole/test_thread.rb
@@ -6,7 +6,7 @@
 require 'test/unit'
 
 if defined?(WIN32OLE)
-  class TestThread < Test::Unit::TestCase
+  class TestWIN32OLE_THREAD < Test::Unit::TestCase
     #
     # test for Bug #2618(ruby-core:27634)
     #
